File: platform/godot/plugin/src/image.cpp
#include "image.h"

#include <godot_cpp/core/class_db.hpp>
#include <godot_cpp/variant/utility_functions.hpp>

#include <cstring>
#include <algorithm>

namespace godot {

// 把 Variant (int 或 [c0(,c1(,c2))]) 规范化为 3 通道 int (不足按末值广播)
static void normalizeRange3(const Variant &p_v, int32_t p_out[3]) {
    if (p_v.get_type() == Variant::ARRAY) {
        Array a = (Array)p_v;
        int n = a.size();
        if (n <= 0) {
            p_out[0] = p_out[1] = p_out[2] = 0;
            return;
        }
        p_out[0] = (int32_t)(int64_t)a[0];
        p_out[1] = (n >= 2) ? (int32_t)(int64_t)a[1] : p_out[0];
        p_out[2] = (n >= 3) ? (int32_t)(int64_t)a[2] : p_out[1];
    } else {
        int32_t iv = (int32_t)(int64_t)p_v;
        p_out[0] = p_out[1] = p_out[2] = iv;
    }
}

AvoxImage::AvoxImage() {}

AvoxImage::~AvoxImage() {
    release();
}

void AvoxImage::release() {
    // delete 路径与 python %newobject 一致 (IImageBuffer 有 virtual ~)
    if (owned && buffer) {
        delete buffer;
    }
    buffer = nullptr;
    owned = false;
}

void AvoxImage::reset_owned() {
    release();
    buffer = avox::createImageBuffer();
    owned = true;
}

// ── 加载/保存 ──

bool AvoxImage::load(const String &p_path) {
    if (p_path.is_empty()) return false;
    reset_owned();
    return avox::loadImagePath(p_path.utf8().get_data(), buffer);
}

bool AvoxImage::load_asset(const String &p_name) {
    if (p_name.is_empty()) return false;
    reset_owned();
    return avox::loadImageAsset(p_name.utf8().get_data(), buffer);
}

bool AvoxImage::save(const String &p_path) {
    if (!buffer || p_path.is_empty()) return false;
    return avox::saveImagePath(p_path.utf8().get_data(), buffer);
}

// ── Godot Image 互转 ──

void AvoxImage::from_image(const Ref<Image> &p_image) {
    if (!p_image.is_valid() || p_image->is_empty()) {
        return;
    }
    // Godot Image 格式不一, 先规整到 RGBA8 (duplicate 不污染调用方)
    Ref<Image> src = p_image;
    if (src->get_format() != Image::FORMAT_RGBA8) {
        src = src->duplicate();
        if (src.is_valid()) {
            src->convert(Image::FORMAT_RGBA8);
        }
    }
    if (!src.is_valid()) return;
    int w = src->get_width();
    int h = src->get_height();
    if (w <= 0 || h <= 0) return;
    reset_owned();
    avox::ImageFormat fmt;
    fmt.width = w;
    fmt.height = h;
    fmt.rowPitch = 0;  // = width * pixelSize (4)
    fmt.imageType = avox::ImageType::rgba8;
    buffer->setImageFormat(fmt);
    int32_t bs = buffer->getBufferSize();
    if (bs <= 0) return;
    PackedByteArray data = src->get_data();
    // RGBA8 level-0 在 get_data() 起始处连续 w*h*4; 有 mipmap 也只拷 level0
    int64_t copyBytes = std::min<int64_t>(bs, (int64_t)data.size());
    if (copyBytes > 0) {
        memcpy(buffer->getPointer(), data.ptr(), (size_t)copyBytes);
    }
}

Ref<Image> AvoxImage::to_image() const {
    if (!buffer) return Ref<Image>();
    avox::ImageFormat fmt = buffer->getImageFormat();
    if (!fmt.bVailid()) return Ref<Image>();
    int32_t w = fmt.width;
    int32_t h = fmt.height;
    const uint8_t *src = buffer->getPointer();
    int32_t bs = buffer->getBufferSize();
    if (!src || bs <= 0 || w <= 0 || h <= 0) return Ref<Image>();
    int32_t pixelSize = avox::getPixelSize(fmt.imageType);
    if (pixelSize <= 0) return Ref<Image>();
    int32_t pitch = (fmt.rowPitch > 0) ? fmt.rowPitch : (w * pixelSize);
    // Godot Image 一律输出 RGBA8 (R,G,B,A); 按 avox imageType 分派转换, 尊重 pitch
    PackedByteArray rgba;
    rgba.resize((int64_t)w * h * 4);
    uint8_t *out = rgba.ptrw();
    int64_t count = (int64_t)w * h;
    switch (fmt.imageType) {
    case avox::ImageType::rgba8:
        if (pitch == w * 4) {
            memcpy(out, src, (size_t)(count * 4));
        } else {
            for (int32_t y = 0; y < h; ++y) {
                memcpy(out + (int64_t)y * w * 4, src + (int64_t)y * pitch, (size_t)w * 4);
            }
        }
        break;
    case avox::ImageType::bgra8:
        // B,G,R,A → R,G,B,A (swap byte0/byte2)
        for (int32_t y = 0; y < h; ++y) {
            const uint8_t *row = src + (int64_t)y * pitch;
            uint8_t *dst = out + (int64_t)y * w * 4;
            for (int32_t x = 0; x < w; ++x) {
                dst[x * 4 + 0] = row[x * 4 + 2];
                dst[x * 4 + 1] = row[x * 4 + 1];
                dst[x * 4 + 2] = row[x * 4 + 0];
                dst[x * 4 + 3] = row[x * 4 + 3];
            }
        }
        break;
    case avox::ImageType::argb8:
        // A,R,G,B → R,G,B,A
        for (int32_t y = 0; y < h; ++y) {
            const uint8_t *row = src + (int64_t)y * pitch;
            uint8_t *dst = out + (int64_t)y * w * 4;
            for (int32_t x = 0; x < w; ++x) {
                dst[x * 4 + 0] = row[x * 4 + 1];
                dst[x * 4 + 1] = row[x * 4 + 2];
                dst[x * 4 + 2] = row[x * 4 + 3];
                dst[x * 4 + 3] = row[x * 4 + 0];
            }
        }
        break;
    case avox::ImageType::rgb8:
        // R,G,B → R,G,B,255
        for (int32_t y = 0; y < h; ++y) {
            const uint8_t *row = src + (int64_t)y * pitch;
            uint8_t *dst = out + (int64_t)y * w * 4;
            for (int32_t x = 0; x < w; ++x) {
                dst[x * 4 + 0] = row[x * 3 + 0];
                dst[x * 4 + 1] = row[x * 3 + 1];
                dst[x * 4 + 2] = row[x * 3 + 2];
                dst[x * 4 + 3] = 255;
            }
        }
        break;
    case avox::ImageType::bgr8:
        // B,G,R → R,G,B,255
        for (int32_t y = 0; y < h; ++y) {
            const uint8_t *row = src + (int64_t)y * pitch;
            uint8_t *dst = out + (int64_t)y * w * 4;
            for (int32_t x = 0; x < w; ++x) {
                dst[x * 4 + 0] = row[x * 3 + 2];
                dst[x * 4 + 1] = row[x * 3 + 1];
                dst[x * 4 + 2] = row[x * 3 + 0];
                dst[x * 4 + 3] = 255;
            }
        }
        break;
    case avox::ImageType::r8:
        // gray → R=G=B=v, A=255
        for (int32_t y = 0; y < h; ++y) {
            const uint8_t *row = src + (int64_t)y * pitch;
            uint8_t *dst = out + (int64_t)y * w * 4;
            for (int32_t x = 0; x < w; ++x) {
                uint8_t v = row[x];
                dst[x * 4 + 0] = v;
                dst[x * 4 + 1] = v;
                dst[x * 4 + 2] = v;
                dst[x * 4 + 3] = 255;
            }
        }
        break;
    default:
        // yuv/float/平面等不在 to_image 直转范围 (留给专用路径)
        return Ref<Image>();
    }
    return Image::create_from_data(w, h, false, Image::FORMAT_RGBA8, rgba);
}

// ── 原始字节 ──

PackedByteArray AvoxImage::to_bytes() const {
    PackedByteArray out;
    if (!buffer) return out;
    int32_t bs = buffer->getBufferSize();
    if (bs <= 0) return out;
    out.resize(bs);
    memcpy(out.ptrw(), buffer->getPointer(), (size_t)bs);
    return out;
}

bool AvoxImage::from_bytes(const PackedByteArray &p_data) {
    if (!buffer) return false;
    int32_t bs = buffer->getBufferSize();
    if (bs <= 0 || p_data.size() < bs) return false;
    memcpy(buffer->getPointer(), p_data.ptr(), (size_t)bs);
    return true;
}

// ── 图像算子 ──

Ref<AvoxImage> AvoxImage::resize(int64_t p_w, int64_t p_h) const {
    if (!buffer || p_w <= 0 || p_h <= 0) return Ref<AvoxImage>();
    Ref<AvoxImage> out = memnew(AvoxImage);
    out->reset_owned();
    if (!avox::resizeImage(buffer, out->buffer, (int32_t)p_w, (int32_t)p_h)) {
        return Ref<AvoxImage>();
    }
    return out;
}

Ref<AvoxImage> AvoxImage::crop(int64_t p_x, int64_t p_y, int64_t p_w, int64_t p_h) const {
    if (!buffer || p_w <= 0 || p_h <= 0) return Ref<AvoxImage>();
    Ref<AvoxImage> out = memnew(AvoxImage);
    out->reset_owned();
    if (!avox::cropImage(buffer, out->buffer, (int32_t)p_x, (int32_t)p_y, (int32_t)p_w, (int32_t)p_h)) {
        return Ref<AvoxImage>();
    }
    return out;
}

Ref<AvoxImage> AvoxImage::to_gray() const {
    if (!buffer) return Ref<AvoxImage>();
    avox::IImageBuffer *g = avox::toGrayImage(buffer);
    if (!g) return Ref<AvoxImage>();
    Ref<AvoxImage> out = memnew(AvoxImage);
    out->buffer = g;
    out->owned = true;
    return out;
}

// ── 编码 ──

String AvoxImage::to_base64(int64_t p_encode_type, int64_t p_quality) const {
    if (!buffer) return String();
    avox::IEncodeConfig cfg;
    // -1 = 默认 jpg (与 python 一致)
    cfg.encodeType = (p_encode_type < 0) ? avox::IEncodeType::jpg : static_cast<avox::IEncodeType>(p_encode_type);
    cfg.quality = (int32_t)p_quality;
    const char *b64 = avox::getImageBase64(buffer, cfg);
    // 静态缓冲, 立即深拷 (见 AvoxVideo.h 注释)
    if (!b64) return String();
    return String::utf8(b64);
}

// ── 像素统计 ──

float AvoxImage::count_in_range(const Variant &p_roi, const Variant &p_lo, const Variant &p_hi) const {
    if (!buffer) return 0.0f;
    avox::ImageFormat fmt = buffer->getImageFormat();
    int32_t x, y, w, h;
    if (p_roi.get_type() == Variant::ARRAY) {
        Array a = (Array)p_roi;
        if (a.size() >= 4) {
            x = (int32_t)(int64_t)a[0];
            y = (int32_t)(int64_t)a[1];
            w = (int32_t)(int64_t)a[2];
            h = (int32_t)(int64_t)a[3];
        } else {
            x = 0;
            y = 0;
            w = fmt.width;
            h = fmt.height;
        }
    } else {
        x = 0;
        y = 0;
        w = fmt.width;
        h = fmt.height;
    }
    int32_t lo3[3], hi3[3];
    normalizeRange3(p_lo, lo3);
    normalizeRange3(p_hi, hi3);
    return avox::countInRange(buffer, x, y, w, h,
                             lo3[0], lo3[1], lo3[2], hi3[0], hi3[1], hi3[2]);
}

// ── 格式/尺寸 ──

int64_t AvoxImage::get_width() const {
    if (!buffer) return 0;
    return buffer->getImageFormat().width;
}

int64_t AvoxImage::get_height() const {
    if (!buffer) return 0;
    return buffer->getImageFormat().height;
}

int64_t AvoxImage::get_image_type() const {
    if (!buffer) return (int64_t)avox::ImageType::other;
    return (int64_t)buffer->getImageFormat().imageType;
}

int64_t AvoxImage::get_buffer_size() const {
    if (!buffer) return 0;
    return buffer->getBufferSize();
}

bool AvoxImage::is_valid() const {
    if (!buffer) return false;
    avox::ImageFormat fmt = buffer->getImageFormat();
    return fmt.bVailid() && buffer->getBufferSize() > 0;
}

bool AvoxImage::set_format(int64_t p_w, int64_t p_h, int64_t p_image_type, int64_t p_row_pitch) {
    if (p_w <= 0 || p_h <= 0) return false;
    reset_owned();
    avox::ImageFormat fmt;
    fmt.width = (int32_t)p_w;
    fmt.height = (int32_t)p_h;
    fmt.rowPitch = (int32_t)p_row_pitch;
    fmt.imageType = static_cast<avox::ImageType>(p_image_type);
    buffer->setImageFormat(fmt);
    return true;
}

// ── 内部 native 互操作 (C++-only, 不绑定 GDScript) ──

Ref<AvoxImage> AvoxImage::from_native_borrowed(avox::IImageBuffer *p_buf) {
    Ref<AvoxImage> img = memnew(AvoxImage);
    img->buffer = p_buf;
    img->owned = false;  // 生命周期归外部 (如 ScreenCapture.getBuffer)
    return img;
}

Ref<AvoxImage> AvoxImage::from_native_owned(avox::IImageBuffer *p_buf) {
    Ref<AvoxImage> img = memnew(AvoxImage);
    img->buffer = p_buf;
    img->owned = true;
    return img;
}

void AvoxImage::_bind_methods() {
    // 加载/保存
    ClassDB::bind_method(D_METHOD("load", "path"), &AvoxImage::load);
    ClassDB::bind_method(D_METHOD("load_asset", "name"), &AvoxImage::load_asset);
    ClassDB::bind_method(D_METHOD("save", "path"), &AvoxImage::save);
    // Godot Image 互转
    ClassDB::bind_method(D_METHOD("from_image", "image"), &AvoxImage::from_image);
    ClassDB::bind_method(D_METHOD("to_image"), &AvoxImage::to_image);
    // 原始字节
    ClassDB::bind_method(D_METHOD("to_bytes"), &AvoxImage::to_bytes);
    ClassDB::bind_method(D_METHOD("from_bytes", "data"), &AvoxImage::from_bytes);
    // 图像算子
    ClassDB::bind_method(D_METHOD("resize", "width", "height"), &AvoxImage::resize);
    ClassDB::bind_method(D_METHOD("crop", "x", "y", "width", "height"), &AvoxImage::crop);
    ClassDB::bind_method(D_METHOD("to_gray"), &AvoxImage::to_gray);
    // 编码
    ClassDB::bind_method(D_METHOD("to_base64", "encode_type", "quality"), &AvoxImage::to_base64, DEFVAL(-1), DEFVAL(85));
    // 像素统计
    ClassDB::bind_method(D_METHOD("count_in_range", "roi", "lo", "hi"), &AvoxImage::count_in_range);
    // 尺寸/格式 getter
    ClassDB::bind_method(D_METHOD("get_width"), &AvoxImage::get_width);
    ClassDB::bind_method(D_METHOD("get_height"), &AvoxImage::get_height);
    ClassDB::bind_method(D_METHOD("get_image_type"), &AvoxImage::get_image_type);
    ClassDB::bind_method(D_METHOD("get_buffer_size"), &AvoxImage::get_buffer_size);
    ClassDB::bind_method(D_METHOD("is_valid"), &AvoxImage::is_valid);
    ClassDB::bind_method(D_METHOD("set_format", "width", "height", "image_type", "row_pitch"), &AvoxImage::set_format, DEFVAL(0));
    // 只读属性 (空 setter = 不可写, 避免引用未绑定 setter 报错)
    ADD_PROPERTY(PropertyInfo(Variant::INT, "width"), "", "get_width");
    ADD_PROPERTY(PropertyInfo(Variant::INT, "height"), "", "get_height");
    ADD_PROPERTY(PropertyInfo(Variant::INT, "image_type"), "", "get_image_type");
}

} // namespace godot

File: swig/python/avox/image.py
# avox SDK Python 高层封装 — Image 模块
# IImageBuffer + 图像加载/保存/缩放/裁剪/base64/bytes + 格式工具

import AvoxWrapper as _pw
from avox._core import ImageType, YuvType, IEncodeType, buffer_to_bytes, bytes_to_buffer

# ─── IImageBuffer ────────────────────────────────────────


class IImageBuffer:
    """图像缓冲封装 (AvoxVideo.h IImageBuffer)。

    持有一块连续像素内存 (BGRA8/RGBA8/RGB8/R8/YUV...), 是图像操作与视觉识别的
    基本数据单元。无参构造会 createImageBuffer() 并由 Python GC 释放 (%newobject);
    传入 native 则为借用 (如 ScreenCapture.getBuffer() 返回的, 生命周期归原主)。
    """

    def __init__(self, native=None):
        # native=None: 自创建自拥有; 否则借用外部 native (不负责释放)
        self._native = native if native else _pw.createImageBuffer()

    # ── 格式 ──

    def setFormat(self, width, height, imageType, rowPitch=0):
        """设置图像格式。imageType 为 ImageType 枚举, rowPitch=0 时默认等于 width*像素大小。"""
        fmt = _pw.ImageFormat()
        fmt.width = width
        fmt.height = height
        fmt.imageType = imageType
        fmt.rowPitch = rowPitch
        self._native.setImageFormat(fmt)
        return self

    def getFormat(self):
        """返回原生 ImageFormat (有 .width/.height/.imageType/.rowPitch)。"""
        return self._native.getImageFormat()

    def bufferSize(self):
        """连续缓冲字节数 (= getPointer() 指向的内存大小)。"""
        return self._native.getBufferSize()

    def bDataRef(self):
        """数据是否引用外部数据 (非自有拷贝)。"""
        return self._native.bDataRef()

    # ── 字节读写 ──

    def to_bytes(self):
        """整块像素拷贝出为 Python bytes (长度 = bufferSize())。"""
        return buffer_to_bytes(self._native)

    def from_bytes(self, data):
        """把 bytes 写入缓冲 (要求 format 已设, 且 len(data)==bufferSize())。"""
        return bytes_to_buffer(self._native, data)

    # ── 拷贝 ──

    def copyFrom(self, src, copyData=True):
        """从 src 拷贝。copyData=True 拷数据, False 只拷 dataptr/size/format。"""
        self._native.copyFrom(src._native if isinstance(src, IImageBuffer) else src, copyData)
        return self

    def copyTo(self, dest, copyData=True):
        """拷贝到 dest。dest 可为 IImageBuffer 或原生对象。"""
        nativeDest = dest._native if isinstance(dest, IImageBuffer) else dest
        self._native.copyTo(nativeDest, copyData)
        return self

    def clear(self):
        """清空 (释放/置零数据, 重置格式)。"""
        self._native.clear()
        return self

    # ── 实例便利 (内部走 free function) ──

    def save(self, path):
        """保存到文件 (按后缀选格式: .png/.jpg/.bmp/.tga, 默认 PNG)。"""
        return _pw.saveImagePath(path, self._native)

    def to_base64(self, encodeType=None, quality=85):
        """返回 Base64 字符串。encodeType 为 IEncodeType, 默认 jpg。"""
        config = _pw.IEncodeConfig()
        if encodeType is not None:
            config.encodeType = encodeType
        config.quality = quality
        return _pw.getImageBase64(self._native, config)

    def resize(self, width, height):
        """缩放, 返回新的 IImageBuffer (失败 None)。不要在 tick 中用。"""
        return resize(self, width, height)

    def crop(self, x, y, width, height):
        """从 (x,y) 裁 width*height, 返回新的 IImageBuffer (失败 None)。不要在 tick 中用。"""
        return crop(self, x, y, width, height)

    def toGray(self):
        """转单通道灰度 (0.299R+0.587G+0.114B), 返回新的 IImageBuffer (r8, 失败 None)。"""
        return toGray(self)

    def countInRange(self, roi, lo, hi):
        """ROI 内 "像素所有通道 ∈ [lo,hi]" 的像素占比 [0,1] (像素级掩码语义, 同 cv::inRange)。
        roi=(x,y,w,h) 或 None=整图; lo/hi 为标量(广播三通道)或 (c0,c1,c2);
        通道序跟随原生格式 (bgr8/bgra8=BGR, rgb8/rgba8=RGB, r8 仅 c0)。"""
        return countInRange(self, roi, lo, hi)

    def saveBinary(self, path):
        """以原始二进制保存 (loadImageBufferBinary 的逆)。"""
        return _pw.saveImageBufferBinary(path, self._native)

    # ── 属性 (从 getImageFormat 取) ──

    @property
    def width(self):
        return self._native.getImageFormat().width

    @property
    def height(self):
        return self._native.getImageFormat().height

    @property
    def imageType(self):
        return self._native.getImageFormat().imageType

    @property
    def rowPitch(self):
        return self._native.getImageFormat().rowPitch

    # ── 生命周期 ──

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.destroy()
        return False

    def destroy(self):
        """释放引用。自创建的由 GC 释放, 这里仅断开 Python 侧引用。"""
        self._native = None


# ─── 模块级工厂 / 图像操作 ───────────────────────────────


def createImageBuffer():
    """创建空 IImageBuffer (随后需 setFormat 或 loadImage 填充)。"""
    return IImageBuffer()


def loadImage(path):
    """从完整路径加载图像 (BMP/PNG/JPG/TGA), 返回 IImageBuffer 或 None。"""
    buf = IImageBuffer()
    if _pw.loadImagePath(path, buf._native):
        return buf
    return None


def loadImageAsset(name):
    """从 assets/images 加载图像 (传文件名, 自动拼路径), 返回 IImageBuffer 或 None。"""
    buf = IImageBuffer()
    if _pw.loadImageAsset(name, buf._native):
        return buf
    return None


def saveImage(path, buf):
    """保存图像到文件 (按后缀选格式)。buf 为 IImageBuffer。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    return _pw.saveImagePath(path, native)


def loadBinary(path):
    """从 saveImageBufferBinary 保存的二进制装载到新 IImageBuffer。"""
    buf = IImageBuffer()
    if _pw.loadImageBufferBinary(path, buf._native):
        return buf
    return None


def resize(buf, width, height):
    """缩放 buf 到 width*height, 返回新 IImageBuffer (失败 None)。不要在 tick 中用。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    out = IImageBuffer()
    if _pw.resizeImage(native, out._native, width, height):
        return out
    return None


def crop(buf, x, y, width, height):
    """从 buf 的 (x,y) 裁 width*height, 返回新 IImageBuffer (失败 None)。不要在 tick 中用。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    out = IImageBuffer()
    if _pw.cropImage(native, out._native, x, y, width, height):
        return out
    return None


def toGray(buf):
    """buf 转单通道灰度 (r8, 0.299R+0.587G+0.114B), 返回新 IImageBuffer (失败 None)。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    out = _pw.toGrayImage(native)
    return IImageBuffer(out) if out else None


def _normRange3(v):
    """标量(广播)或序列 -> [c0,c1,c2] int。"""
    if isinstance(v, (list, tuple)):
        a = [int(x) for x in v]
        while len(a) < 3:
            a.append(a[-1])
        return a[:3]
    iv = int(v)
    return [iv, iv, iv]


def countInRange(buf, roi, lo, hi):
    """ROI 内 "像素所有通道 ∈ [lo,hi]" 的像素占比 [0,1] (像素级掩码语义, 同 cv::inRange)。
    roi=(x,y,w,h) 或 None=整图; lo/hi 为标量(广播三通道)或 (c0,c1,c2);
    通道序跟随原生格式 (bgr8/bgra8=BGR, rgb8/rgba8=RGB, r8 仅 c0)。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    if roi is None:
        fmt = native.getImageFormat()
        x, y, w, h = 0, 0, fmt.width, fmt.height
    else:
        x, y, w, h = int(roi[0]), int(roi[1]), int(roi[2]), int(roi[3])
    lo3 = _normRange3(lo)
    hi3 = _normRange3(hi)
    return _pw.countInRange(native, x, y, w, h,
                            lo3[0], lo3[1], lo3[2], hi3[0], hi3[1], hi3[2])


def toBase64(buf, encodeType=None, quality=85):
    """返回 buf 的 Base64 字符串。encodeType 为 IEncodeType, 默认 jpg。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    config = _pw.IEncodeConfig()
    if encodeType is not None:
        config.encodeType = encodeType
    config.quality = quality
    return _pw.getImageBase64(native, config)


def toBytes(buf):
    """buf 整块像素 → bytes。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    return buffer_to_bytes(native)


def fromBytes(buf, data):
    """bytes → buf (format 需已设, len(data)==bufferSize())。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    return bytes_to_buffer(native, data)


# ─── 格式 / 尺寸工具 ──────────────────────────────────────


def getPixelSize(imageType):
    """ImageType 单像素字节数。"""
    return _pw.getPixelSize(imageType)


def getImageSize(imageFormat):
    """ImageFormat 的缓冲字节数。imageFormat 可为原生 ImageFormat。"""
    return _pw.getImageSize(imageFormat)


def getImageTypeStr(imageType):
    """ImageType → 字符串。"""
    return _pw.getImageTypeStr(imageType)


def getYuvTypeStr(yuvType):
    """YuvType → 字符串。"""
    return _pw.getYuvTypeStr(yuvType)


def bVPlaneFormat(yuvType):
    """是否平面 YUV 格式 (420P/422P/444P/NV12)。"""
    return _pw.bVPlaneFormat(yuvType)


# ─── 进阶 YUV 转换 ──────────────────────────────────────
# 链路1: ISurfaceRender.onFrame(YUVFrame) → yuvframe2Rgba(frame, buf) → IImageBuffer(RGBA)
# 链路2: IImageBuffer(YUV) --image2YUVFrame--> YUVFrame --yuvframe2Rgba--> IImageBuffer(RGBA)
# unpackGpuYUV: GPU 渲染输出的 YUV IImageBuffer, UV padding 重排给 FFmpeg 读 (原地)
# YUVFrame.data/stride 由 C++ 填 (onFrame 回调 / image2YUVFrame), Python 不手填。

def getYuvFrameSize(yuvFormat, rowPitch):
    """计算 YUV 帧字节数。yuvFormat 为 _pw.YUVFormat。"""
    return _pw.getYuvFrameSize(yuvFormat, rowPitch)


def image2YUVFrame(buf, yuvType):
    """从 IImageBuffer (R8) 导出 _pw.YUVFrame; 返回 (ok, yuvFrame)。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    frame = _pw.YUVFrame()
    ok = _pw.image2YUVFrame(native, frame, yuvType)
    return ok, frame


def yuvframe2Rgba(frame, buf):
    """YUVFrame → IImageBuffer(RGBA)。
    frame 为 _pw.YUVFrame (常用 image2YUVFrame 导出的, data 指向某 IImageBuffer 内部);
    buf 为目标 IImageBuffer (会被内部 setFormat 为 width×height rgba8)。
    返回 bool。仅支持 yuv420P / nv12。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    return _pw.yuvframe2Rgba(frame, native)


def unpackGpuYUV(buf, yuvType):
    """GPU 渲染输出 YUV 的 UV padding 重排 (原地, 不可逆)。
    buf 为含 YUV 平面数据的 IImageBuffer (rowPitch != width 时生效);
    yuvType 为 YuvType (yuv420P/yuv422P)。重排后 stride=rowPitch/2, FFmpeg 等距读取。"""
    native = buf._native if isinstance(buf, IImageBuffer) else buf
    _pw.unpackGpuYUV(native, yuvType)
